src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/haflow/sync/actions/EmitSyncRuleRequestsAction.java

```diff
@@ -0,0 +1,129 @@
+public class EmitSyncRuleRequestsAction extends HaFlowRuleManagerProcessingAction<
+        HaFlowSyncFsm, State, Event, HaFlowSyncContext> {
+
+    public EmitSyncRuleRequestsAction(PersistenceManager persistenceManager, RuleManager ruleManager) {
+        super(persistenceManager, ruleManager);
+    }
+
+    @Override
+    protected void perform(State from, State to, Event event, HaFlowSyncContext context, HaFlowSyncFsm stateMachine) {
+        Collection<InstallSpeakerCommandsRequest> installRequests = buildSpeakerRequests(stateMachine);
+
+        if (installRequests.isEmpty()) {
+            stateMachine.saveActionToHistory("No requests to sync HA-flow rules");
+            stateMachine.fire(Event.SKIP_RULES_SYNC);
+        } else {
+            // emitting
+            installRequests.forEach(request -> {
+                stateMachine.getCarrier().sendSpeakerRequest(request);
+                stateMachine.addSpeakerCommand(request.getCommandId(), request);
+                stateMachine.addPendingCommand(request.getCommandId(), request.getSwitchId());
+            });
+
+            stateMachine.saveActionToHistory("Commands for syncing HA-flow rules have been sent");
+        }
+    }
+
+    private List<SpeakerData> buildRules(HaFlowPath haPath) {
+        Set<SwitchId> switchIds = haPath.getAllInvolvedSwitches();
+        Set<PathId> pathIds = new HashSet<>(haPath.getSubPathIds());
+        DataAdapter dataAdapter = new PersistenceDataAdapter(persistenceManager, pathIds, switchIds, false);
+        return ruleManager.buildRulesHaFlowPath(haPath, true, dataAdapter);
+    }
+
+    private Collection<HaFlowPath> getPaths(String haFlowId) {
+        HaFlow haFlow = getHaFlow(haFlowId);
+        Collection<HaFlowPath> paths = haFlow.getPrimaryPaths();
+
+        if (haFlow.getProtectedForwardPath() != null && haFlow.getProtectedReversePath() != null) {
+            paths.add(haFlow.getProtectedForwardPath());
+            paths.add(haFlow.getProtectedReversePath());
+        }
+        return paths;
+    }
+
+    private Collection<InstallSpeakerCommandsRequest> buildSpeakerRequests(HaFlowSyncFsm stateMachine) {
+        Collection<InstallSpeakerCommandsRequest> result = new ArrayList<>();
+        for (HaFlowPath haFlowPath : getPaths(stateMachine.getHaFlowId())) {
+            List<SpeakerData> speakerCommands = buildRules(haFlowPath);
+            Collection<InstallSpeakerCommandsRequest> installRequests = buildHaFlowInstallRequests(
+                    speakerCommands, stateMachine.getCommandContext(), false);
+
+            for (InstallSpeakerCommandsRequest installRequest : installRequests) {
+                result.add(installRequest);
+                Map<SwitchId, Set<PathId>> subPathMap = buildSubPathBySwitchIdMap(haFlowPath.getSubPaths());
+
+                stateMachine.putPathRequest(installRequest.getCommandId(), haFlowPath.getHaPathId());
+                for (PathId subPathId : subPathMap.get(installRequest.getSwitchId())) {
```

Comment:
Please, rewrite into smallest methods. Triple nested `for` loop is very hard to read (and, eventually, modify).